### PR TITLE
ANDAUTH-34 Improve logic of contacts synchonization

### PR DESCRIPTION
**Pull request by InsanusMokrassar**
* rewrite XWikiHttp and SyncAdapter to work via subjects

### CODE REVIEW

File: app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java [hunk 1]
```diff
@@ -98,22 +100,51 @@ public void call(Throwable throwable) {
...
+                            } else {
+                                throw new IOException(TAG + "syncType error, SyncType=" + syncType);
+                            }
+                        } catch (IOException e) {
+                            subject.onError(e);
+                        }
+                    }
+                }
+            ).start();
+            semaphore.acquire();
```
Comment (pa1pal) [inline app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java:143]: why this `aquire()` at the end?

File: app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java [hunk 2]
```diff
@@ -98,22 +100,51 @@ public void call(Throwable throwable) {
...
      * @throws XmlPullParserException
      */
-    public static SyncData getSyncData(String lastSyncTime, int syncType) throws IOException, XmlPullParserException {
-        if (syncType == Constants.SYNC_TYPE_ALL_USERS) {
-            return getSyncAllUsers(lastSyncTime);
-        } else if (syncType == Constants.SYNC_TYPE_SELECTED_GROUPS) {
-            List<String> groupIdList = SharedPrefsUtils.getArrayList(AppContext.getInstance().getApplicationContext(), Constants.SELECTED_GROUPS);
-            return getSyncGroups(groupIdList, lastSyncTime);
+    public static Observable<XWikiUserFull> getSyncData(final int syncType) {
+        final PublishSubject<XWikiUserFull> subject = PublishSubject.create();
```
Comment (pa1pal) [inline app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java:111]: Shouldn't we use `ReplaySubject` here, it works but I think might be we lose some contacts in the beginning.

Comment (InsanusMokrassar) [reply app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java:111]: `ReplySubject` will store all data which was called by `onNext`. It is not better than using old `SyncData`.

File: app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java (hunk 1 above)
Comment (InsanusMokrassar) [reply app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java:143]: Here we are waiting moment when thread will be ready to wait notification from subject.

Comment (InsanusMokrassar) [reply app/src/main/java/org/xwiki/android/authenticator/rest/XWikiHttp.java:143]: [Here](https://github.com/xwiki-contrib/android-authenticator/blob/4d115e6f23a938eb84f21ef5ef7dd237b491dbef/app/src/main/java/org/xwiki/android/authenticator/syncadapter/SyncAdapter.java#L133) we notify that all actions with `subscription` was done

**Review by pa1pal — APPROVED** [review]

**Pull request merged**